content: Implement Pygments Monokai, for upcoming dark theme

chrisbobbe opened the pull request:
For dark theme, web currently uses a mix of Pygments Monokai and Pygments Default. Mixing in Pygments Default seems accidental, so here we just use pure Pygments Monokai. [Discussion](https://chat.zulip.org/#narrow/stream/431-redesign-project/topic/code.20span.20colors/near/1832283).

I've included screenshots, below, of just Pygments Monokai, and of Pygments Monokai with web's adjustments. (Except for the adjustment of putting a border around `err` text; that's hard in Flutter.)

Compare with Pygments's own demonstration of Pygments Monokai here: https://pygments.org/styles/

(To make these screenshots, I plugged in dark-theme colors for the code-block background and message background. So just focus on the code blocks and disregard the jarring appearance of light-theme-styled elements elsewhere. When we implement dark theme #95, we'll do it coherently.)

| This PR (pure Pygments Monokai) | Following web (having some Pygments Default mixed in) |
| --- | --- |
| ![8F5F5901-6293-4CF6-9A16-13516A8C2813](https://github.com/zulip/zulip-flutter/assets/22248748/99cbacae-e37c-42ec-a2fe-76589b751654) | ![D497DCF5-8115-469B-BBCF-C6827DF33DEE](https://github.com/zulip/zulip-flutter/assets/22248748/9c690dd3-e705-4b24-9679-04a84457c4df) |

Fixes: #749

gnprice commented on lib/widgets/code_block.dart:
```diff
@@ -31,18 +32,34 @@ class CodeBlockTextStyles {
       // .err { border: 1px solid hsl(0deg 100% 50%); }
       err: const TextStyle(backgroundColor: Color(0xffe2706e)),
 
+      esc: null,
```
While we're at it, write the TODO(#754) to update the light-theme
    styles. (Hopefully to something that also has a style for all of
    `CodeBlockSpanType`'s values.)

This change is fine, but there's nothing wrong with some of the values being null — that just means those spans inherit the base styling from `plain`, right?

I believe the empty-string values in the upstream file:
https://github.com/pygments/pygments/blob/f64833d9d/pygments/styles/monokai.py
mean inheriting the style, so e.g. the classes `w` and `x`, and `g` and `gr`, all inherit the base style. (While e.g. `cm` inherits that of `c`.)

chrisbobbe replied:
`plain` copies a variable in web, `--color-markdown-code-text` (starting at this PR's commit

```
content: Follow web in code-block default text color
```

). I don't think that variable or its light/dark values has anything to do with Pygments styles; I think it's mostly meant for plain un-syntax-highlighted code blocks.

From a Pygments [doc](https://pygments.org/docs/styledevelopment/) on writing your own custom style, I think the "base style" that `w`, `x`, `g`, and `gr` are meant to inherit from is actually the [`Token` style](https://github.com/pygments/pygments/blob/f64833d9d/pygments/styles/monokai.py#L32), which for Monokai is `#f8f8f2`. Here's from that doc:

> `Token` is the default style inherited by all token types.

This probably explains why web gives `color: #f8f8f2;` for `w`, `x`, `g`, and `gr`, instead of letting them fall back to `--color-markdown-code-text`. In zulip/zulip@f020f9eee, the switch to Monokai, it says the CSS was generated from a CLI tool from Pygments.

My process in making this commit was to follow web (minus some overrides; [discussion](https://chat.zulip.org/#narrow/stream/431-redesign-project/topic/code.20span.20colors/near/1832283)) and then check the result against that upstream monokai.py file. So that's why our `w`, `x`, `g`, and `gr` are `0xfff8f8f2` instead of white with 85% opacity, and I think that's correct. Maybe

chrisbobbe replied:
So I guess having gone through all this just now, I'm now questioning some of our lines in the `switch (type)`:

```dart
CodeBlockSpanType.text => null, // A span with type of text is always unstyled.
```

```dart
CodeBlockSpanType.unknown => null,
```

gnprice replied:
Yeah, I agree with that conclusion.

gnprice replied:
Not a blocker for merging this, though — that'd be a fine followup. I don't think this PR causes any regressions on that front.

chrisbobbe replied:
Actually it's a bit more complicated. In the Pygments HTML, text that is meant to get the `Text` style seems to be represented without `<span>`s—which is the same way text appears in plain, un-Pygmented code blocks. For example, this:

```diff
diff --git lib/widgets/content.dart lib/widgets/content.dart
index 1715d41e7..43c747232 100644
--- lib/widgets/content.dart
+++ lib/widgets/content.dart
@@ -37,6 +37,7 @@ class ContentTheme extends ThemeExtension<ContentTheme> {
       colorCodeBlockBackground: const HSLColor.fromAHSL(0.04, 0, 0, 0).toColor(),
       colorDirectMentionBackground: const HSLColor.fromAHSL(0.2, 240, 0.7, 0.7).toColor(),
       colorGlobalTimeBackground: const HSLColor.fromAHSL(1, 0, 0, 0.93).toColor(),
+      colorGlobalTimeBorder: const HSLColor.fromAHSL(1, 0, 0, 0.8).toColor(),
       colorMathBlockBorder: const HSLColor.fromAHSL(0.15, 240, 0.8, 0.5).toColor(),
       colorMessageMediaContainerBackground: const Color.fromRGBO(0, 0, 0, 0.03),
       colorThematicBreak: const HSLColor.fromAHSL(1, 0, 0, .87).toColor(),
@@ -61,6 +62,7 @@ class ContentTheme extends ThemeExtension<ContentTheme> {
       colorCodeBlockBackground: const HSLColor.fromAHSL(0.04, 0, 0, 1).toColor(),
       colorDirectMentionBackground: const HSLColor.fromAHSL(0.25, 240, 0.52, 0.6).toColor(),
       colorGlobalTimeBackground: const HSLColor.fromAHSL(0.2, 0, 0, 0).toColor(),
+      colorGlobalTimeBorder: const HSLColor.fromAHSL(0.4, 0, 0, 0).toColor(),
       colorMathBlockBorder: const HSLColor.fromAHSL(1, 240, 0.4, 0.4).toColor(),
       colorMessageMediaContainerBackground: const HSLColor.fromAHSL(0.03, 0, 0, 1).toColor(),
       colorThematicBreak: const HSLColor.fromAHSL(1, 0, 0, .87).toColor().withOpacity(0.2),
```

The diff-untouched lines are just text-node children of the `<code>`, not wrapped in spans. (They start with a `.w` span with a single space, not sure why.)

<details>

```html
<p><span class="user-mention" data-user-id="13313">@Chris Bobbe</span></p>
<p><span class="user-mention" data-user-id="17327">@Chris Bobbe (Test Account)</span></p><p><span class="user-mention" data-user-id="17327">@Chris Bobbe (Test Account)</span></p><div class="spoiler-block"><div class="spoiler-header">
<p>regular <strong>bold</strong></p>
</div><div class="spoiler-content" aria-hidden="true">
<p>content</p>
</div></div><h1><del><strong>bold</strong></del></h1><div class="spoiler-block"><div class="spoiler-header">
<p><em>italic <strong>bold</strong></em></p>
</div><div class="spoiler-content" aria-hidden="true">
<p>content</p>
</div></div><h1><span class="user-mention silent" data-user-id="13313">Chris Bobbe</span></h1><div class="spoiler-block"><div class="spoiler-header">
<h1>h1: regular <strong>bold</strong></h1>
</div><div class="spoiler-content" aria-hidden="true">
<p>content</p>
</div></div><div class="spoiler-block"><div class="spoiler-header">
<h1>h1</h1>
</div><div class="spoiler-content" aria-hidden="true">
<p>content</p>
</div></div><div class="codehilite" data-code-language="Dart"><pre><span></span><code><span class="kd">class</span><span class="w"> </span><span class="nc">ContentExample</span><span class="w"> </span><span class="p">{</span>
</code></pre></div><div class="codehilite" data-code-language="Diff"><pre><span></span><code><span class="gh">diff --git lib/widgets/content.dart lib/widgets/content.dart</span>
<span class="gh">index 1715d41e7..43c747232 100644</span>
<span class="gd">--- lib/widgets/content.dart</span>
<span class="gi">+++ lib/widgets/content.dart</span>
<span class="gu">@@ -37,6 +37,7 @@ class ContentTheme extends ThemeExtension&lt;ContentTheme&gt; {</span>
<span class="w"> </span>      colorCodeBlockBackground: const HSLColor.fromAHSL(0.04, 0, 0, 0).toColor(),
<span class="w"> </span>      colorDirectMentionBackground: const HSLColor.fromAHSL(0.2, 240, 0.7, 0.7).toColor(),
<span class="w"> </span>      colorGlobalTimeBackground: const HSLColor.fromAHSL(1, 0, 0, 0.93).toColor(),
<span class="gi">+      colorGlobalTimeBorder: const HSLColor.fromAHSL(1, 0, 0, 0.8).toColor(),</span>
<span class="w"> </span>      colorMathBlockBorder: const HSLColor.fromAHSL(0.15, 240, 0.8, 0.5).toColor(),
<span class="w"> </span>      colorMessageMediaContainerBackground: const Color.fromRGBO(0, 0, 0, 0.03),
<span class="w"> </span>      colorThematicBreak: const HSLColor.fromAHSL(1, 0, 0, .87).toColor(),
<span class="gu">@@ -61,6 +62,7 @@ class ContentTheme extends ThemeExtension&lt;ContentTheme&gt; {</span>
<span class="w"> </span>      colorCodeBlockBackground: const HSLColor.fromAHSL(0.04, 0, 0, 1).toColor(),
<span class="w"> </span>      colorDirectMentionBackground: const HSLColor.fromAHSL(0.25, 240, 0.52, 0.6).toColor(),
<span class="w"> </span>      colorGlobalTimeBackground: const HSLColor.fromAHSL(0.2, 0, 0, 0).toColor(),
<span class="gi">+      colorGlobalTimeBorder: const HSLColor.fromAHSL(0.4, 0, 0, 0).toColor(),</span>
<span class="w"> </span>      colorMathBlockBorder: const HSLColor.fromAHSL(1, 240, 0.4, 0.4).toColor(),
<span class="w"> </span>      colorMessageMediaContainerBackground: const HSLColor.fromAHSL(0.03, 0, 0, 1).toColor(),
<span class="w"> </span>      colorThematicBreak: const HSLColor.fromAHSL(1, 0, 0, .87).toColor().withOpacity(0.2),
</code></pre></div>
```

</details>

That's the same way "foo" appears in an unstyled code block:

```
foo
```

```html
<div class="codehilite"><pre><span></span><code>foo
</code></pre></div>
```

So when we see text like that, we'll want to apply `Text` / `Token` style if it's a syntax-highlighted block (and I guess ditto for unknown tokens), and Zulip's own `--color-markdown-code-text` otherwise.

Web doesn't seem to handle this wrinkle. With a color meter, I can see the diff-untouched lines in the highlighted block are the same color as "foo" in the unhighlighted block. Link to those messages: https://chat.zulip.org/#narrow/stream/7-test-here/topic/content/near/1841644

chrisbobbe replied:
I think we can check if a code block is syntax-highlighted by looking for a `data-code-language` in the `codehilite.zulip-code-block` div.

chrisbobbe replied:
Actually, you know what, it's probably `--color-markdown-code-text` that we should ignore, and just use the Pygments `Text` / `Token` color wherever we were using that. That'll be way simpler in code.

gnprice reviewed (commented):
Thanks! These changes all LGTM. One commit-message comment below.

Can you also say a few words about how you produced the list of styles? The link in the comment is to upstream, but then the comments on individual styles are in terms of CSS, which that upstream file isn't.

> Oh, and I owe another pair of screenshots; ~coming soon~ here they are. This PR includes a commit to change light mode's plain code-block text color (used for un-highlighted code blocks) from `hsl(0deg 0% 15%)` to true black, following web.

Does this have a visible effect in the screenshots? I flipped back and forth, and couldn't spot any difference at all except in the clock at the top of the screen.

chrisbobbe commented:
> > Oh, and I owe another pair of screenshots; ~coming soon~ here they are. This PR includes a commit to change light mode's plain code-block text color (used for un-highlighted code blocks) from `hsl(0deg 0% 15%)` to true black, following web.
> 
> Does this have a visible effect in the screenshots? I flipped back and forth, and couldn't spot any difference at all except in the clock at the top of the screen.

It's very subtle. The plain, unstyled code block text is now true black, instead of `hsl(0deg 0% 15%)` i.e. `ContentTheme.textStylePlainParagraph.color`. Do you have something like Mac's "Digital Color Meter" to see the difference? That's how I checked that there definitely was a change, but I'm also able to see a small difference visually on my monitor.

chrisbobbe commented:
> Can you also say a few words about how you produced the list of styles?

I looked at web's `%dark-theme` block in web/styles/pygments.css, bulk-formatted it to make nice single-line comments like `// .ss { color: #e6db74; }`, and translated the commented CSS into Flutter `TextStyle`s.

(Then I started applying the "non-color Zulip customizations" from `pygments.css` before realizing we didn't want those, as mentioned, and I removed what I had added.)

gnprice commented:
> It's very subtle. The plain, unstyled code block text is now true black, instead of `hsl(0deg 0% 15%)` i.e. `ContentTheme.textStylePlainParagraph.color`.

Oh wow, OK — yeah, if I bring my face right up a few inches from the screen and *then* flip back and forth, I do see the difference.

gnprice commented:
> I looked at web's `%dark-theme` block in web/styles/pygments.css, bulk-formatted it to make nice single-line comments like `// .ss { color: #e6db74; }`, copy-pasted those into this file, and translated the commented CSS into Flutter `TextStyle`s.

Cool, makes sense. I think a few words saying these came from web/styles/pygments.css like that would be helpful.

Maybe just have this comment:
```dart
  // Pygments Monokai:
  //   https://github.com/pygments/pygments/blob/f64833d9d/pygments/styles/monokai.py
  factory CodeBlockTextStyles.dark(BuildContext context) {
```

mention that, in addition to the upstream Pygments link (that the current revision has) for whence the colors were ultimately derived.

chrisbobbe commented:
Thanks for the review! Revision pushed, and I've made a note about the followup you mentioned: https://github.com/zulip/zulip-flutter/pull/756#discussion_r1653830011

chrisbobbe commented:
Just resolved the conflicts from #759

gnprice commented:
Thanks! Looks good; merging.
